executor/window.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/executor/windowfuncs"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// WindowExec is the executor for window functions.
+type WindowExec struct {
+	baseExecutor
+
+	groupChecker  *groupChecker
+	inputIter     *chunk.Iterator4Chunk
+	inputRow      chunk.Row
+	groupRows     []chunk.Row
+	childResults  []*chunk.Chunk
+	windowFunc    windowfuncs.WindowFunc
+	partialResult windowfuncs.PartialResult
+	executed      bool
+	meetNewGroup  bool
+}
+
+// Close implements the Executor Close interface.
+func (e *WindowExec) Close() error {
+	e.childResults = nil
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+// Next implements the Executor Next interface.
+func (e *WindowExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("windowExec.Next", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+	}
+	if e.runtimeStats != nil {
+		start := time.Now()
+		defer func() { e.runtimeStats.Record(time.Now().Sub(start), chk.NumRows()) }()
+	}
+	chk.Reset()
+	if e.meetNewGroup && e.windowFunc.HasRemainingResults() {
+		err := e.appendResult2Chunk(chk)
+		if err != nil {
+			return err
+		}
+	}
+	for !e.executed && (chk.NumRows() == 0 || chk.RemainedRows(chk.NumCols()-1) > 0) {
+		err := e.consumeOneGroup(ctx, chk)
+		if err != nil {
+			e.executed = true
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *WindowExec) consumeOneGroup(ctx context.Context, chk *chunk.Chunk) error {
+	var err error
+	if err = e.fetchChildIfNecessary(ctx, chk); err != nil {
+		return errors.Trace(err)
+	}
+	for ; e.inputRow != e.inputIter.End(); e.inputRow = e.inputIter.Next() {
+		e.meetNewGroup, err = e.groupChecker.meetNewGroup(e.inputRow)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if e.meetNewGroup {
+			err := e.consumeGroupRows(chk)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = e.appendResult2Chunk(chk)
```

Comment:
After consuming group rows, we may buffered many input `Chunk`s, so `appendResult2Chunk()` should be called multiple times?

Reply:
Yes, it will. In https://github.com/pingcap/tidb/pull/8899/files#diff-509643ad918102b310cd4be4c55fc644R58, we will check if the last group is finished and has remaining rows, and if so, it will be called.